File: neetcode150/stack/leetcode22.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
        To use a stack with this, we essentially perform some sort of BFS

        :param n:
        :return:
        """
        result = []
        stack = [("(", 1, 0)]

        while stack:
            cur, opn, close = stack.pop()
            if opn < close or opn > n or close > n:
                continue

            if opn == n and close == n:
                result.append(cur)

            stack.append((cur + "(", opn + 1, close))
            stack.append((cur + ")", opn, close + 1))

        return result

File: js_algorithms/neetcode150/stack/leetcode22.js
/**
 * @param {number} n
 * @return {string[]}
 */
const generateParenthesis = function (n) {
  const result = [];
  const stack = [];
  stack.push({ path: '', open: 0, close: 0 });

  while (stack.length) {
    const { path, open, close } = stack.pop();

    if (path.length === 2 * n) {
      result.push(path);
    } else {
      if (open < n) {
        stack.push({ path: `${path}(`, open: open + 1, close });
      }

      if (close < open) {
        stack.push({ path: `${path})`, open, close: close + 1 });
      }
    }
  }

  return result;
};

module.exports = generateParenthesis;
